### PR TITLE
Merge TRP3 Function and String Interpolation Utilities

### PR DESCRIPTION
**Pull request by Meorawr**
Based off the `feature/pooled_class` branch due to the dependency upon that.

This adds `Strings.interpolate(formatString: string, replacements: table)` to provide a more powerful `string.format` that includes support for named replacements in the form `%KEY$specifier`.

```lua
local result = Strings.interpolate("Elsa is %ELSA_AGE$03d years old.", { ELSA_AGE = 21 });
-- Result: "Elsa is 021 years old."
```

There's also a new `Funcs` utility area for function utilities, right now there's just `Funcs.bind(fn: function, value: any)` which creates a closure that'll forward `value` to `fn` as the first argument, followed by the arguments to the closure itself.

```lua
local function praise(name)
    print(name .. " is the best damn Disney princess of all time.");
end

local bound = Funcs.bind(praise, "Elsa");
bound();
-- Outputs: "Elsa is the best damn Disney princess of all time."
```

### CODE REVIEW

**Comment by Ellypse** [comment]
Be careful of the forces you are meddling with. I will not tolerate that insults of the sorts are made against the best Disney princess of all times.

**Comment by Meorawr** [comment]
You're right of course. The examples have been updated accordingly.

I also now have "elsa frozen age" in my search history, but this is a non-issue.

File: Tools/Funcs.lua [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+end
+
+-- Lua imports
+local assert = assert;
+local setmetatable = setmetatable;
+local type = type;
+local unpack = unpack;
+
+-- Ellyb imports
+local Funcs = Ellyb.Funcs;
```
Comment (Ellypse) [inline Tools/Funcs.lua]: This smells like a copy-pasta error 🤔

File: Tools/Funcs.lua [hunk 2]
```diff
@@ -0,0 +1,33 @@
+---@type Ellyb
+local Ellyb = Ellyb(...);
+
+if Ellyb.Funcs then
+	return
+end
+
+-- Lua imports
+local assert = assert;
```
Comment (Ellypse) [inline Tools/Funcs.lua]: You should remove unused imports

**Comment by Ellypse** [comment]
Open question: for now the library's modules have names like `ColorManager` instead of `ColorMgr` or `Strings` instead of `Str`. What do you think of having your `Funcs` module be called `Functions` (the name with an uppercase is not reserved)?

**Comment by Meorawr** [comment]
That last commit should tell you exactly what I think of the idea. Not sure why I picked Funcs in the first place...

Rebased to fix up the totally-not-copy-pasted upvalues.

**Comment by Meorawr** [comment]
I've replaced `PooledClass` with a mixin since that was my original plan for that utility anyway.

The existing implementation wasn't very generic and would have broken the instant someone subclassed a class that inherited `PooledClass` because subclasses would be put into the same pool as the parent class, so instantiating either could randomly give you parent or child class instances.

The mixin approach is better since you can just attach it to any class without imposing constraints on the hierarchy (you don't need class A to inherit `PooledClass` if you want subclass B to be poolable), and correctly handles the pooling of subclass instances by segmenting them into tables based on their class.

...Bloody hell that's a lot of "class".

**Pull request merged**